### PR TITLE
sorter/leveldb(ticdc): optimize write and read throughput

### PR DESCRIPTION
**Pull request by overvenus**
### What problem does this PR solve?

This PR

* optimize write throughput by switch leveldb to pebble.
* optimize write throughput by fully asynchronously read and write in Poll.
* optimize read throughput by reducing calling `Iterator.First()` which may take up to minutes to return.

| master | this PR |
| -- | -- |
| image | image | 

Issue Number: Ref https://github.com/pingcap/tiflow/issues/3227

### Check List 

Tests 

 - Unit test
 - Integration test

Side effects

 - Increased code complexity
   - Mix up asynchronously read and write in the same function is complex, we need to refactor Poll into read/write actor model in the future.

Related changes

 - Need to cherry-pick to the release branch

### Release note 

```release-note
None
```

### CODE REVIEW

File: cdc/sorter/leveldb/table_sorter.go [hunk 1]
```diff
@@ -304,54 +322,75 @@ func (ls *Sorter) outputBufferedResolvedEvents(
...
+			commitTs = event.CRTs
 		}
 		// Group resolved events that has the same commit ts.
-		if lastCommitTs == event.CRTs {
+		if commitTs == event.CRTs {
 			buffer.appendResolvedEvent(event)
 			continue
 		}
-		// Output buffered events. The current event belongs to a new group.
+		// As new event belongs to a new txn group, we need to output
```
Comment (asddongmen) [inline cdc/sorter/leveldb/table_sorter.go:377]: "As a new event ..."

File: cdc/sorter/leveldb/table_sorter.go [hunk 2]
```diff
@@ -304,54 +322,75 @@ func (ls *Sorter) outputBufferedResolvedEvents(
...
+		// all buffered events before append the event.
 		ls.outputBufferedResolvedEvents(buffer, true)
 		lenResolvedEvents, _ = buffer.len()
 		if lenResolvedEvents > 0 {
-			// Output blocked, break and free iterator.
+			// Output blocked, skip append new event.
+			// This means we has not read next.
```
Comment (asddongmen) [inline cdc/sorter/leveldb/table_sorter.go:383]: "This means we have not read Next"

**Review by asddongmen — APPROVED** [review]

File: cdc/sorter/leveldb/table_sorter.go [hunk 3]
```diff
@@ -304,54 +322,75 @@ func (ls *Sorter) outputBufferedResolvedEvents(
...
-// to send them later.
+// later, and appends resolved events to outputBuffer resolvedEvents to send
+// them later.
 // outputBuffer must be empty.
 func (ls *Sorter) outputIterEvents(
-	iter db.Iterator, buffer *outputBuffer, maxResolvedTs uint64,
-) (uint64, error) {
+	iter db.Iterator, hasReadLastNext bool, buffer *outputBuffer,
+	resolvedTs uint64,
+) (bool, uint64, error) {
```
Comment (amyangfei) [inline cdc/sorter/leveldb/table_sorter.go:331]: Please describe the meaning of return variable

File: cdc/sorter/leveldb/table_sorter.go [hunk 4]
```diff
@@ -474,62 +605,46 @@ func (ls *Sorter) poll(ctx context.Context, state *pollState) error {
...
-	// Send write task to leveldb.
+	var hasIter bool
+	task.IterReq, hasIter = state.tryGetIterator(ls.uid, ls.tableID)
+	// Send write/read task to leveldb.
 	err = ls.router.SendB(ctx, ls.actorID, actormsg.SorterMessage(task))
-	if err != nil {
+	if err != nil || !hasIter {
+		// Skip read iterator if send fails or there is no iterator.
 		return errors.Trace(err)
 	}
```
Comment (amyangfei) [inline cdc/sorter/leveldb/table_sorter.go:635]: `err == nil && !hasIter` should be another error, instead of `nil`

**Review by amyangfei — APPROVED** [review]

**Comment by lonng** [comment]
Is there any benchmark results of comparing improvement?

**Comment by overvenus** [comment]
Updated in PR description

**Pull request merged**